### PR TITLE
Adds test which pulls dockerhub Windows image

### PR DESCRIPTION
**Pull request by claudiubelu**
**What type of PR is this?**

/kind failing-test
/sig testing
/sig windows

**What this PR does / why we need it**:

Updates the test "should be able to pull image from docker hub" to pull
the golang image instead of alpine, which can also be used on Windows,
allowing the test to pass on both OSes.

**Which issue(s) this PR fixes**:

Related: #69871

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dims** [comment]
@bclau problem is that it will not match the description ` "should be able to pull image from docker hub"`

**Comment by bclau** [comment]
@dims Well, the busybox image is pulled from the "dockerLibraryRegistry"[1], so it would still be on docker hub.

The image used should be configurable, somehow, since it is OS dependent. Not only that, but for Windows Server 2019, we cannot run containers with images based on Windows Server 1803 and vice-versa, so I couldn't simply hardcode an image name like "microsoft/windowservercore:1803" in a test.

[1] https://github.com/kubernetes/kubernetes/blob/master/test/utils/image/manifest.go#L99

**Comment by PatrickLang** [comment]
@dims since upstream busybox doesn't publish Windows images, we could pick another image. 

https://hub.docker.com/_/golang

On Windows 10 (with Docker for Windows - Linux mode)
```
docker version
Client: Docker Engine - Community
 Version:           18.09.0
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        4d60db4
 Built:             Wed Nov  7 00:47:51 2018
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       4d60db4
  Built:            Wed Nov  7 00:55:00 2018
  OS/Arch:          linux/amd64
  Experimental:     true

C:\Users\plang\Source\enhancements\keps\sig-windows>docker pull golang
Using default tag: latest
latest: Pulling from library/golang
cd8eada9c7bb: Pull complete
c2677faec825: Pull complete
fcce419a96b1: Pull complete
045b51e26e75: Pull complete
70a7003e9fe9: Pull complete
751b6f060321: Pull complete
462a907acb44: Pull complete
Digest: sha256:0c7ba0b5eff462068cb8c16d94248ed44463c30418fac91ee46283e34f483547
Status: Downloaded newer image for golang:latest
```

On Windows Server 2019
```
docker pull golang
Using default tag: latest
latest: Pulling from library/golang
65014b3c3121: Already exists
2ac060f1ef06: Already exists
409ca3908f9b: Pull complete
05c8e42e8c40: Pull complete
f39532532c4d: Pull complete
0faa97f71daa: Pull complete
5a99d3cca2b4: Pull complete
75bc32531793: Pull complete
6e7d996a4bf1: Pull complete
3e248d960d69: Pull complete
87723d90b848: Pull complete
0851883a4fe9: Pull complete
fcf9ba83111c: Pull complete
Digest: sha256:0c7ba0b5eff462068cb8c16d94248ed44463c30418fac91ee46283e34f483547
Status: Downloaded newer image for golang:latest

C:\Users\vagrant>docker version
Client:
 Version:           18.09.1
 API version:       1.39
 Go version:        go1.10.6
 Git commit:        20b67756d0
 Built:             unknown-buildtime
 OS/Arch:           windows/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.1
  API version:      1.39 (minimum version 1.24)
  Go version:       go1.10.6
  Git commit:       20b67756d0
  Built:            01/09/2019 17:09:57
  OS/Arch:          windows/amd64
  Experimental:     false
```

**Comment by PatrickLang** [comment]
@bclau wouldn't `ping -t localhost` be ok on Linux too?

**Comment by dims** [comment]
@PatrickLang you mean `ping -t 1 localhost`? (-t is the TTL)

**Comment by PatrickLang** [comment]
@dims Hmm, no. `-t` on windows makes it keep pinging until it gets a ctrl-c. Looks like that won't work on both platforms

**Comment by bclau** [comment]
Indeed. I couldn't find a proper command which does the same thing and be valid on both OSes.

**Comment by benmoss** [comment]
`curl --retry 20 --retry-connrefused --retry-delay 900 localhost:9999` looks like it should suffice if all you need is a Linux & Windows cross-platform "runs (more or less) forever" command

**Comment by PatrickLang** [comment]
ah yes good idea @benmoss 

One caveat - be sure not to do this if you're using a Windows image with the shell set to PowerShell. It has an alias from `curl` to `Invoke-WebRequest` which has different syntax. You can get around that by doing `curl.exe` but that would break Linux.

**Comment by michmike** [comment]
is there anyone that still has objections on this PR and the implementation of ping or the choice of the image?
adding a hold, but also approving it in the meantime
/hold

**Review by michmike — APPROVED** [review]

**Comment by bclau** [comment]
@benmoss Nice, thanks for the command! It works fine on Windows Server 2019. I'm not sure curl exists in Windows Server 1803, but we've decided to go forth with WS 2019 anyways.

Indeed, curl is aliased in powershell to Invoke-WebRequest, but most images use cmd.exe as the entrypoint, so as long as that doesn't change, it should be fine.

**Comment by benmoss** [comment]
@bclau yeah curl was added in 1803 https://blogs.msdn.microsoft.com/commandline/2018/03/07/windows10v1803/

if i had any sway in Microsoft land I would highly petition for deprecating the curl Powershell alias 😄

**Review by neolit123 — APPROVED** [review]
/skip
/lgtm
/priority backlog

**Comment by BenTheElder** [comment]
How much bigger is this? Ideally we keep the image small to reduce flake chances.

**Comment by bclau** [comment]
@BenTheElder Apparently, the golang image is ~80 MB, looking on dockerhub: [1], while the alpine image is ~2-3 MB [2]. AFAIK, they're the compressed sizes. Still, for downloading, 80 MB is reasonable IMO.

[1] https://hub.docker.com/_/golang?tab=tags
[2] https://hub.docker.com/_/alpine?tab=tags

**Comment by neolit123** [comment]
> @BenTheElder Apparently, the golang image is ~80 MB, looking on dockerhub: [1], while the alpine image is ~2-3 MB [2]. AFAIK, they're the compressed sizes. Still, for downloading, 80 MB is reasonable IMO.

just downloaded the image locally and it's 816MB

$ docker pull golang:latest
$ docker image inspect golang:latest --format='{{.Size}}'
816133367

can we figure out a different image?
or install tools on top of the minimal alpine image?

also `latest` is not an ideal tag.

/lgtm cancel

File: test/e2e/common/pods.go
Comment (BenTheElder) [inline test/e2e/common/pods.go]: Why did this change?

**Comment by benmoss** [comment]
If we already have to have a conditional around the test itself, why not just make a conditional around the image selection?

**Comment by michmike** [comment]
@bclau can you look into a couple of things
1. is there a different golang image that might be smaller in size
2. what are some of the actual commands that need to pass both on windows and linux and this golang image supports


@BenTheElder and @neolit123 it is possible that we may not be able to find a smaller image that satisfies  #2. what's the net effect impact of a large image in tests beyond extending the testing time

File: test/e2e/common/pods.go
Comment (neolit123) [inline test/e2e/common/pods.go]: if `golang:latest` is 800mb, 15 seconds might not be enough for a non-flaky test.

File: test/e2e/common/runtime.go
Comment (BenTheElder) [inline test/e2e/common/runtime.go]: Agree with @neolit123, can we not use a :latest at least?
Ideally we want to pull the smallest viable image as well.

**Comment by bclau** [comment]
Looking for a smaller image. At the moment, the smallest I found that also has sleep (needed by the test) is ``plugins/git``, which has 62.2MB.

File: test/e2e/common/pods.go
Comment (bclau) [reply test/e2e/common/pods.go]: My bad. I was testing something and had this local change and I didn't realize I amended with this change.

**Comment by neolit123** [comment]
@bclau 

sleep is already available in `alpine:3.7`
```
docker run -it alpine:3.7 /bin/sh -c "sleep --help"
BusyBox v1.27.2 (2018-06-06 09:08:44 UTC) multi-call binary.
```
are there more tools that are needed?

**Comment by bclau** [comment]
The best I've found would be ``plugins/base``, it has 5 MB, it has sleep, which is needed for the test. Works on Windows and Linux.

https://hub.docker.com/r/plugins/manifest

@neolit123 We can't use alpine on Windows.

**Comment by neolit123** [comment]
> @neolit123 We can't use alpine on Windows.

`plugins/base` is `alpine:3.8` based as a Linux container
https://github.com/drone-plugins/drone-base/blob/master/docker/Dockerfile.linux.amd64

but nano-server based as Windows container:
https://github.com/drone-plugins/drone-base/blob/master/docker/Dockerfile.windows.1809

is this container supposed to run as a native one?
(i'm going to assume yes as per the `ping -t` call.
in which case the size would be +200MB at least.

i think we need to make the images conditional as @benmoss suggested earlier.
so fix linux to `alpine:3.7` and use another container on Windows. `plugins/base` is not ideal because it's unversioned.

**Comment by michmike** [comment]
> > @neolit123 We can't use alpine on Windows.
> 
> `plugins/base` is `alpine:3.8` based as a Linux container
> https://github.com/drone-plugins/drone-base/blob/master/docker/Dockerfile.linux.amd64
> 
> but nano-server based as Windows container:
> https://github.com/drone-plugins/drone-base/blob/master/docker/Dockerfile.windows.1809
> 
> is this container supposed to run as a native one?
> (i'm going to assume yes as per the `ping -t` call.
> in which case the size would be +200MB at least.
> 
> i think we need to make the images conditional as @benmoss suggested earlier.
> so fix linux to `alpine:3.7` and use another container on Windows? `plugins/base` is not ideal because it's unversioned.

i agree with making images conditional. we may solve the issue today by finding a universal image but long term this is not a viable strategy. also i would rather a high percentage of the tests to complete quickly with a small image and only the windows tests to experience a small slowdown.

Now for the big question:
1. it is easy to make the images conditional for a test that specifically tests linux or windows but not both. How do we make images conditional based on the node that ends up being the target of the test. This is going to be specifically important when we test heterogeneous clusters with both linux and windows compute nodes.

**Comment by benmoss** [comment]
> Now for the big question:
> 
> 1. it is easy to make the images conditional for a test that specifically tests linux or windows but not both. How do we make images conditional based on the node that ends up being the target of the test. This is going to be specifically important when we test heterogeneous clusters with both linux and windows compute nodes.

I think that's pretty out of scope for this PR. I think we're going to need global test flags that specify whether you intend to test your cluster's Linux or Windows nodes

**Comment by neolit123** [comment]
@benmoss 

> I think we're going to need global test flags that specify whether you intend to test your cluster's Linux 
or Windows nodes

isn't framework.NodeOSDistroIs sufficient?

@bclau 

> So, probably the most foolproof plan for this test would be to leave it as is, and create a separate Windows-specific test which will use an image with a manifest list.
>
> The reason for the manifest list is because at the moment, Container's OS version must match the host's OS version, otherwise the container won't start, so using a manifest list will make it more future-proof.

is this the related explanation?
https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility#windows-server-containers

but this is about Windows Server containers.
are we going to run Windows Server based containers only in these tests?

the manifest list JSON stays behind an endpoint such as `myimage:mytag`
from there the schema.v2 can handle the host matching based on `platform` -> `os.version`.
so this can technically can be done in the same file using framework.NodeOSDistroIs and no new file would be needed. unless i'm missing something?

are there existing images that will satisfy the manifest list requirement - e.g. at docker hub?

**Comment by PatrickLang** [comment]
https://hub.docker.com/_/hello-world?tab=tags may be a smaller image - 134MB on Windows Server 1809. Still need to check if it has the needed manifest list

**Comment by neolit123** [comment]
@BenTheElder i participated in a SIG Windows meeting and for the Windows side of this runtime test the container OS version has to match the host OS version. this forces us to download at least 100 MB.

so we have two options here:
- use such a 100 MB image for Windows and use the existing alpine:3.7 on Linux (i.e. abstract either in go or schema.v2), possibly extend timeouts where needed.
- don't run this test on Windows due to the chances for flakes.

WDYT?

**Comment by BenTheElder** [comment]
Thanks, the first option SGTM, if Windows experiences flakes we can revisit later / punt to SIG-Windows, and in the meantime we won't regress on Linux.

The second option seems a bit overkill, we should at least get it working on Windows and go from there.

**Comment by michmike** [comment]
any update on this? We can use the image Patrick mentioned (134mbs) and also option 1 from Lubomir

**Comment by neolit123** [comment]
@bclau let me know if you don't have the time for this i can try sending a new PR.

**Comment by bclau** [comment]
Ok, so, plenty to respond to.

> 
> 
> @benmoss
> 
> > I think we're going to need global test flags that specify whether you intend to test your cluster's Linux
> > or Windows nodes
> 
> isn't framework.NodeOSDistroIs sufficient?
> 

It is. We have some tests which are already using that.

> @bclau
> 
> > So, probably the most foolproof plan for this test would be to leave it as is, and create a separate Windows-specific test which will use an image with a manifest list.
> > The reason for the manifest list is because at the moment, Container's OS version must match the host's OS version, otherwise the container won't start, so using a manifest list will make it more future-proof.
> 
> is this the related explanation?
> https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility#windows-server-containers
> 
> but this is about Windows Server containers.
> are we going to run Windows Server based containers only in these tests?
> 

Indeed it is, and yes, we are using Windows Server nodes (+ Docker SE) for testing. I don't think there's a need to test Docker CE, and that wouldn't be the focus right now anyways.

> the manifest list JSON stays behind an endpoint such as `myimage:mytag`
> from there the schema.v2 can handle the host matching based on `platform` -> `os.version`.
> so technically this can be done in the same file using framework.NodeOSDistroIs and no new file would be needed. unless i'm missing something?
> 
> are there existing images that will satisfy the manifest list requirement - e.g. at docker hub or do we have to create our own?

We could go with the ``hello-world`` image Patrick suggested. It is a manifest list ( https://paste.ubuntu.com/p/q2rxdhPRr5/ ) and it covers pretty much every OS which is of interest. Regarding its size, 100 is not much, and if it is based on the nanoserver image (which is the base for all kubeletwin/pause images used on Windows nodes), then the total pulled data is less than 100 MB. I've tried this out on a Windows Server 2019 node, and the 1st layer already existed. It only pulled ~31 MB of data.

Keep in mind that the ``hello-world`` Linux image does not have sleep, so we can't use it for both Windows and Linux tests.

IMO, we could just add a separate test which has the "[WindowsOnly]" tag in its name, and we'll only run it if ``--node-os-distro=windows``, so we won't break every other CI with this change.

If you agree with this option, I could send an update with it in the morning.

**Comment by neolit123** [comment]
> Keep in mind that the `hello-world` Linux image does not have sleep, so we can't use it for both Windows and Linux tests.

also `hello-world` is unversioned which isn't ideal.
let's keep `alpine:3.7` for linux in any case.

> IMO, we could just add a separate test which has the "[WindowsOnly]" tag in its name, and we'll only run it if `--node-os-distro=windows`, so we won't break every other CI with this change.

but it means this this test we are trying to fix with this PR has to be [LinuxOnly]?

why not have `framework.NodeOSDistroIs` in this `.go` file that uses a different image for windows? i think this is easier. it's essentially the same test with different image and command on Windows.

in terms of manifest lists for the variety of Windows versions we need to support isn't there an official nanoserver image that is a manifest list? using `hello-world` because it's a manifest list feels odd. we also have the option to build our own and push to docker hub.

**Comment by bclau** [comment]
> 
> 
> > Keep in mind that the `hello-world` Linux image does not have sleep, so we can't use it for both Windows and Linux tests.
> 
> also `hello-world` is unversioned which isn't ideal.
> let's keep `alpine:3.7` for linux in any case.
> 

True, it's unversioned, but it's been the "same" image since its inception and it will continue to be. There is no real reason to change that image or what it does, and I think it's safe to use.

> > IMO, we could just add a separate test which has the "[WindowsOnly]" tag in its name, and we'll only run it if `--node-os-distro=windows`, so we won't break every other CI with this change.
> 
> but it means this test we are trying to fix with this PR has to be [LinuxOnly]?
> 

Well, yeah.

> why not have `framework.NodeOSDistroIs` in this `.go` file that uses a different image for windows? i think this is easier. it's essentially the same test with different image and command on Windows.
> 

I have thought about this, and I had a implementation that did this. But, it wasn't ideal, it would only fix the dockerhub scenario. We should also consider having the gcr.io scenario work with Windows as well (I didn't found a single Windows image there, so we'll probably have to add one). There might also be other image repositories we'll have add tests for in the future. So, separate tests would be the simplest to do and maintain.

> in terms of manifest lists for the variety of Windows versions we need to support isn't there an official nanoserver image that is a manifest list? using `hello-world` because it's a manifest list feels odd. we also have the option to build our own and push to docker hub.

Ehh.. the oficial nanoserver "images" are manifest lists, but per architecture [1]. So, for each Windows version, there's a separate tag (1709, 1803, 1809, etc.), so that's not very futureproof.

And indeed, we could just create and set our own image. For example, we could have a "e2eteam/busybox:1.29" image as a manifest list and use that. And since it's the busybox image, we could use the same command as the Linux image. And as far as size goes, that image is used in tests, so it should already be pre-pulled.

Right now, I'm trying out images with manifest lists. All the images for used for testing under the "claudiubelu" repo are manifest lists (ATM, for WS 1803 and WS 1809), and if everything looks fine, we'll migrate them to the "e2eteam" repo, so "e2eteam/busybox:1.29" would be a viable option for this test.

[1] https://paste.ubuntu.com/p/HgFSbNQc7g/

**Comment by neolit123** [comment]
> I have thought about this, and I had a implementation that did this. But, it wasn't ideal, it would only fix the dockerhub scenario. We should also consider having the gcr.io scenario work with Windows as well (I didn't found a single Windows image there, so we'll probably have to add one). There might also be other image repositories we'll have add tests for in the future. So, separate tests would be the simplest to do and maintain.

so it seems we need to push an image to GCR anyhow.

as far as scenarios go you can have a shared list at the start and append OS specific test cases to the list of test cases depending on `framework.NodeOSDistroIs`

> And indeed, we could just create and set our own image. For example, we could have a "e2eteam/busybox:1.29" image as a manifest list and use that. And since it's the busybox image, we could use the same command as the Linux image. And as far as size goes, that image is used in tests, so it should already be pre-pulled.

ok, sounds good.

**Comment by bclau** [comment]
So, if we are to add a test with our own image which has manifest lists, it would look something like this. Ofc, the repo name will be changed to "e2eteam" after we've verified that they're good and after they've been pushed to the "e2eteam" repo.

File: test/e2e/common/runtime.go
Comment (neolit123) [inline test/e2e/common/runtime.go]: so i guess we need to hold this PR until:
- we can have this image in a official account somewhere?
- add an [WindowsOnly] alternative to the test case "should be able to pull image from gcr.io [LinuxOnly]","

is this correct?

Comment (bclau) [reply test/e2e/common/runtime.go]: Regarding the official account, the ``e2eteam`` repo is what we've created to be used by CIs and they are actively maintained and kept up-to-date, so I think that's a good enough.

Regarding the 2nd point, We could push the busybox image to gcr.io as well, I can look into it, but that could also be done in a separate PR.

Updated the repo name to ``e2eteam``.

Comment (neolit123) [reply test/e2e/common/runtime.go]: > that could also be done in a separate PR.

ok, agreed.

could you please add TODO about the GCR test for Windows?

**Comment by michmike** [comment]
it looks like we are close to finalizing this. great job @bclau . i will approve this with a hold so that you can remove the hold when ready to merge. please make sure to create the second PR for pushing the image to gcr.io.

/hold

**Review by michmike — APPROVED** [review]

**Comment by neolit123** [comment]
/assign @timothysc @vishh 
we can get an approve here please.

this also needs a follow up PR to support GCR too.

File: test/e2e/common/runtime.go
Comment (bclau) [reply test/e2e/common/runtime.go]: Done.

**Comment by bclau** [comment]
So, it seems that storing and pulling images from gcr.io is not quite free [0], so I didn't upload any image. I am assuming that someone is sponsoring some gcr.io repos, since there are 2 repos used in E2E tests (gcr.io/kubernetes-e2e-test-images, gcr.io/k8s-authenticated-test). Any idea how to upload a manifest list there, or who could?

[0] https://cloud.google.com/storage/pricing

**Comment by neolit123** [comment]
/lgtm
on the latest change.

> I didn't upload any image. I am assuming that someone is sponsoring some gcr.io repos

@bclau gcr.io hosting is sponsored by google, at the moment. we need to coordinate the push to the registry with maintainers on slack.

**Review by timothysc — APPROVED** [review]
/approve

**Pull request merged**
